### PR TITLE
Supported parsing date time text with valid zero zone offsets

### PR DESCRIPTION
**Pull request by kevinjom**
Refer to https://github.com/FasterXML/jackson-modules-java8/issues/18

### CODE REVIEW

File: datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/InstantDeserializer.java [hunk 1]
```diff
@@ -51,13 +51,21 @@
 {
     private static final long serialVersionUID = 1L;
 
+    /**
+     * Constants used to check if the time offset is zero. See [jackson-modules-java8#18]
+     *
+     * @since 2.9.0
+     */
+    private static final String ISO8601_UTC_ZERO_OFFSET_SUFFIX_REGEX = "\\+00:?(00)?$";
+    private static final String ISO8601_UTC_ZERO_OFFSET_REGEX = "[^\\+]+" + ISO8601_UTC_ZERO_OFFSET_SUFFIX_REGEX;
```
Comment (cowtowncoder) [inline datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/InstantDeserializer.java]: This should be pre-compiled as `Pattern` for performance reasons.

File: datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/InstantDeserializer.java [hunk 2]
```diff
@@ -285,6 +288,20 @@ private ZoneId getZone(DeserializationContext context)
         return (_valueClass == Instant.class) ? null : context.getTimeZone().toZoneId();
     }
 
+    private String replaceZeroOffsetAsZIfNecessary(String text)
+    {
+        if (replaceZeroOffsetAsZ && endsWithZeroOffset(text)) {
```
Comment (cowtowncoder) [inline datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/InstantDeserializer.java]: This seems to evaluate regexp twice, first to match, then to replace: would it be possible to make that just one operation? Regexp evaluation has non-trivial cost, even considering that textual date handling itself is rather expensive operation.

**Comment by cowtowncoder** [comment]
Thank you for contributing this, yes, makes sense. The only suggestion I have is that since regexp handling has some overhead, it'd be good to do the usual optimization: pre-compile regexp String as `Pattern` (as they are immutable and can be shared, reused), and try to avoid matching and replacing as separate steps.

**Comment by cowtowncoder** [comment]
One other thing: I'm happy to merge this (ideally with changes although I can probably make them myself too if need be), but unless I've asked before we need a filled Contributor License Agreement (CLA) before the first contribution (but just once, it's good for any other patches).
Document is at:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

and usually it's easiest to print, fill & sign, scan, email to `info` at fasterxml dot com.
Once this occurs (apologies if I've asked for this before), I can merge the PR.

Thank you again for the contribution!

File: datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/InstantDeserializer.java (hunk 2 above)
Comment (kevinjom) [inline datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/InstantDeserializer.java]: Thanks for the comment @cowtowncoder,  I just checked the doc, your point makes perfect sense, I was wrong, I remember if no match found that would throw an exception. will fix

File: datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/InstantDeserializer.java (hunk 1 above)
Comment (kevinjom) [inline datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/deser/InstantDeserializer.java]: Yes, will fix

**Comment by kevinjom** [comment]
@cowtowncoder I am really grateful for the comments and excited to contribute to this project, I will sign that CLA and send email later (well, before I fixed all the issues you pointed)

**Comment by kevinjom** [comment]
@cowtowncoder comments fixed and CLA mail sent. Please have a look.

**Comment by cowtowncoder** [comment]
@kevinjom Thank you!

**Pull request merged**